pkg/models/customalerting/rules/utils.go

```diff
@@ -0,0 +1,422 @@
+// into the prometheus query expression, which will limit the query scope.
+func InjectExprNamespaceLabel(expr, namespace string) (string, error) {
+	parsedExpr, err := parser.ParseExpr(expr)
+	if err != nil {
+		return "", err
+	}
+	if err = injectproxy.NewEnforcer(&promlabels.Matcher{
+		Type:  promlabels.MatchEqual,
+		Name:  "namespace",
+		Value: namespace,
+	}).EnforceNode(parsedExpr); err == nil {
+		return parsedExpr.String(), nil
+	}
+	return "", err
+}
+
+func FormatDuration(for_ string) (string, error) {
+	var duration prommodel.Duration
+	var err error
+	if for_ != "" {
+		duration, err = prommodel.ParseDuration(for_)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to parse Duration string(\"%s\") to time.Duration", for_)
+		}
+	}
+	return duration.String(), nil
+}
+
+func parseDurationSeconds(durationSeconds float64) string {
+	return prommodel.Duration(int64(durationSeconds * float64(time.Second))).String()
+}
+
+func GenResourceRuleIdIgnoreFormat(group string, rule *promresourcesv1.Rule) string {
+	query, err := FormatExpr(rule.Expr.String())
+	if err != nil {
+		klog.Warning(errors.Wrapf(err, "invalid alerting rule(%s)", rule.Alert))
+		query = rule.Expr.String()
+	}
+	duration, err := FormatDuration(rule.For)
+	if err != nil {
+		klog.Warning(errors.Wrapf(err, "invalid alerting rule(%s)", rule.Alert))
+		duration = rule.For
+	}
+
+	lbls := make(map[string]string)
+	for k, v := range rule.Labels {
+		lbls[k] = v
+	}
+	lbls[LabelKeyInternalRuleGroup] = group
+	lbls[LabelKeyInternalRuleName] = rule.Alert
+	lbls[LabelKeyInternalRuleQuery] = query
+	lbls[LabelKeyInternalRuleDuration] = duration
+
+	return prommodel.Fingerprint(prommodel.LabelsToSignature(lbls)).String()
+}
+
+func GenEndpointRuleId(group string, epRule *customalerting.AlertingRule,
+	externalLabels func() map[string]string) (string, error) {
+	query, err := FormatExpr(epRule.Query)
+	if err != nil {
+		return "", err
+	}
+	duration := parseDurationSeconds(epRule.Duration)
+
+	var labelsMap map[string]string
+	if externalLabels == nil {
+		labelsMap = epRule.Labels
+	} else {
+		labelsMap = make(map[string]string)
+		extLabels := externalLabels()
+		for key, value := range epRule.Labels {
+			if v, ok := extLabels[key]; !(ok && value == v) {
+				labelsMap[key] = value
+			}
+		}
+	}
+
+	lbls := make(map[string]string)
+	for k, v := range labelsMap {
+		lbls[k] = v
+	}
+	lbls[LabelKeyInternalRuleGroup] = group
+	lbls[LabelKeyInternalRuleName] = epRule.Name
+	lbls[LabelKeyInternalRuleQuery] = query
+	lbls[LabelKeyInternalRuleDuration] = duration
+
+	return prommodel.Fingerprint(prommodel.LabelsToSignature(lbls)).String(), nil
+}
+
+// MixAlertingRules mix rules from prometheusrule custom resources and rules from endpoints.
+// Use rules from prometheusrule custom resources as the main reference.
+func MixAlertingRules(ruleNamespace string, ruleChunk *ResourceRuleChunk, epRuleGroups []*customalerting.RuleGroup,
```

Comment:
to be more clear, maybe we can rename MixAlertingRules/MixAlertingRule/mixAlertingRule to GetAlertingRulesStatus/GetAlertingRuleStatus/getAlertingRuleStatus ?